orttraining/orttraining/eager/ort_aten.cpp

```diff
@@ -1176,32 +1222,284 @@ at::Tensor& mm_out(
-      &at::native::cpu_fallback,
-      ATEN_OP(squeeze)>::call(self);
+        &at::native::cpu_fallback,
+        ATEN_OP(squeeze)>::call(self);
 
   auto& invoker = GetORTInvoker(self.device());
 
   auto ort_input_0_self = create_ort_value(invoker, self);
 
   std::vector<OrtValue> ort_outputs_0_Squeeze(1);
 
-  auto status = invoker.Invoke("Squeeze", {
-    std::move(ort_input_0_self),
-  }, ort_outputs_0_Squeeze, nullptr);
+  auto status = invoker.Invoke("Squeeze",
+                               {std::move(ort_input_0_self)},
+                               ort_outputs_0_Squeeze, nullptr);
   CHECK_STATUS(status);
 
   at::TensorOptions tensor_options = self.options();
   return aten_tensor_from_ort(
-    std::move(ort_outputs_0_Squeeze[0]),
-    tensor_options);
+      std::move(ort_outputs_0_Squeeze[0]),
+      tensor_options);
+}
+
+// aten::add.out(Tensor self, Tensor other, *, Scalar alpha=1, Tensor(a!) out) -> Tensor(a!)
+at::Tensor& add_out(
+    const at::Tensor& self,
+    const at::Tensor& other,
+    // *,
+    const at::Scalar& alpha,
+    at::Tensor& out) {
+  ORT_LOG_FN(self, other, alpha, out);
+
+  auto promoted_type = PromoteScalarTypesWithCategory({other.scalar_type(), self.scalar_type()}, {alpha.type()});
+
+  if (
+      std::vector<at::ScalarType> supportedTypes =
+          {at::kBFloat16, at::kByte, at::kDouble, at::kFloat, at::kHalf, at::kInt, at::kLong, at::kShort};
+      !IsSupportedType(alpha, supportedTypes) ||
+      !IsSupportedType(other, supportedTypes) ||
+      !IsSupportedType(self, supportedTypes) ||
+      !c10::canCast(*promoted_type, out.scalar_type())) {
+    return at::native::call_fallback_fn<
+        &at::native::cpu_fallback,
+        ATEN_OP(add_out)>::call(self, other, alpha, out);
+  }
+  auto& invoker = GetORTInvoker(self.device());
+
+  auto ort_input_0_alpha = create_ort_value(invoker, alpha);
+  if (alpha.type() != *promoted_type) {
+    ort_input_0_alpha = CastToType(invoker, ort_input_0_alpha, *promoted_type);
+  }
+  auto ort_input_0_other = create_ort_value(invoker, other);
+  if (other.scalar_type() != *promoted_type) {
+    ort_input_0_other = CastToType(invoker, ort_input_0_other, *promoted_type);
+  }
+
+  std::vector<OrtValue> ort_outputs_0_Mul(1);
+
+  auto status = invoker.Invoke("Mul",
+                               {std::move(ort_input_0_alpha), std::move(ort_input_0_other)},
+                               ort_outputs_0_Mul, nullptr);
+  CHECK_STATUS(status);
+
+  auto ort_input_1_self = create_ort_value(invoker, self);
+  if (self.scalar_type() != *promoted_type) {
+    ort_input_1_self = CastToType(invoker, ort_input_1_self, *promoted_type);
+  }
+
+  // resize the output and then create output ort value to be updated.
+  onnxruntime::TensorShape out_shape;
+  resize_output(
+      invoker,
+      dynamic_cast<ORTTensorImpl*>(out.unsafeGetTensorImpl()),
+      BroadcastShape("add_out", ort_input_1_self, ort_input_0_other, out_shape));
```

Comment:
shouldn't the 2nd param be `ort_outputs_0_Mul[0]`?

Reply:
Done. Though technically they should be the same shape, but I think your suggest improves the logical walking through of the code.